### ISSUE TITLE
“openssl cms” with EdDSA does not work

### ISSUE THREAD

**Issue by wangweij** [openssl/openssl#11915]
After creating an Ed25519 self-signed certificate with

    openssl req -x509 -subj "/CN=A" -newkey ed25519 -nodes -keyout - > signer

I tried to sign a file with

    openssl cms -in file -sign -signer signer -noattr

but it failed with

    4622915008:error:2E066080:CMS routines:CMS_add1_signer:no default digest:crypto/cms/cms_sd.c:295:

My understanding is the pureEdDSA does not need a message digest at all.

**Comment by levitte** [comment]
What OpenSSL version?

> My understanding is the pureEdDSA does not need a message digest at all.

True.  @DDvO, do you know some more about this?  Is OpenSSL's CMS up to date with RFC 8419?

**Comment by wangweij** [comment]
It's 1.1.1g installed by brew on macOS.

**Comment by DDvO** [comment]
> > My understanding is the pureEdDSA does not need a message digest at all.
> 
> True. @DDvO, do you know some more about this? Is OpenSSL's CMS up to date with RFC 8419?

I don't have particular knowledge of CMS and EdDSA, but will have a look.

Trying to follow @wangweij's steps I've just come across a further issue:
With OpenSSL 1.1.1, 
```
openssl req -x509 -subj "/CN=A" -newkey ed25519 -nodes -keyout - > signer.pem
```
outputs both the key and the cert to `signer.pem`,
but with the current master the cert is not included!
Yet when I replace `ed25519` by `RSA:1024`:
```
apps/openssl req -x509 -subj "/CN=A" -newkey RSA:1024 -nodes -keyout - > signer.pem
```
it works as expected.

**Comment by wangweij** [comment]
I did tried to download/master the master to see if this issue was fixed and noticed that problem. I can confirm that even with the key and cert created from 1.1.1g, the current master shows the same problem at `cms -sign`.

**Comment by DDvO** [comment]
> the current master shows the same problem at cms -sign.

I found this, too.

It looks like `openssl cms -sign` simply does not support pureEdDSA yet.

`EVP_PKEY_get_default_digest_nid()` correctly returns `NID_undef` in this case,
and the CMS app reports this as an error rather than getting along without a digest algorithm.

**Comment by wangweij** [comment]
I tried use `-md sha512` but also not accepted.

**Comment by richsalz** [comment]
> It looks like openssl cms -sign simply does not support pureEdDSA yet

It  does not. Nor does dgst.  I don't think anything does, except calling the low-level EVP routines directly.

**Comment by henning-krause** [comment]
I'm getting the same error with OpenSSL 3.0.0-beta1-dev  (Library: OpenSSL 3.0.0-beta1-dev ). Apparently, this issue hasn't been fixed.

**Comment by DDvO** [comment]
Ping @openssl/committers

**Comment by DDvO** [comment]
This should be solved as specified in https://datatracker.ietf.org/doc/html/rfc8419#section-2.3:

>  When the signer includes signed attributes, a message digest
   algorithm is used to compute the message digest on the eContent
   value.  When signing with Ed25519, the message digest algorithm MUST
   be SHA-512 [FIPS180].  Additional information on SHA-512 is available
   in [RFC6234].  When signing with Ed448, the message digest algorithm
   MUST be SHAKE256 [FIPS202] with a 512-bit output value.

**Comment by DDvO** [comment]
There are clearly bugs here.
Already loading the key and cert supplied via the `-signer` option produces the above error message.
If the key and cert are provided separately, the error is not shown.
In both cases then the process continues but no signature is produced, failing without giving an error.
I am analyzing further and will provide a fix.

**Comment by t8m** [comment]
> There are clearly bugs here.

But we might simply say - supporting EdDSA properly in CMS is a feature. And the bug would be at most that we do not reject EdDSA outright but instead fail in weird ways.

It really depends on the size and nature of the fixes that need to be done for the EdDSA support.

**Comment by t8m** [comment]
It also isn't a regression from 1.1.1 so even if regarded as bugs it could be postponed Post 3.0.

**Comment by slontis** [comment]
I tend to agree that this is really a new feature and could be documented as not currently supported.

**Comment by DDvO** [comment]
So I'm going to split the solution into bug fixes (to be included into 3.0 as far as possble)
and a feature extension (post 3.0).

**Comment by DDvO** [comment]
Due to a request/report yesterday on openssl-users, I got reminded of this issue, which also holds for PKCS7 (e.g, `openssl smime`).

This issue is much harder to fix than for CMP, due to ugly legacy code in the CMS and PKCS7 implementations.
I tried again today, but sorry to say: fixing this is beyond my OpenSSL expertise.

**Comment by DDvO** [comment]
BTW, for the use case given above in the OP, meanwhile for some reason no error queue output is given at all.

In `cms_sd.c`, after 
`if (EVP_PKEY_get_default_digest_nid(pk, &def_nid) <= 0)`
a call
`    ERR_raise(ERR_LIB_CMS, CMS_R_NO_DEFAULT_DIGEST);`
 is missing.

**Comment by vdukhovni** [comment]
> > BTW, for the use case given above in the OP, meanwhile for some reason
> > no error queue output is given at all.

Yes, just an exit error code of "3" with no stderr output.

> > In "cms_sd.c", after 
> >
> >     if (EVP_PKEY_get_default_digest_nid(pk, &def_nid) <= 0)
> >
> > a call
> >
> >      ERR_raise(ERR_LIB_CMS, CMS_R_NO_DEFAULT_DIGEST);
> >
> >  is missing.

A succinct PR to add an error message would be productive.

**Comment by DDvO** [comment]
I investigated somewhat further, and it looks like there is a relatively elegant solution at least for Ed25519 with CMS,
yet some several CMS cases fail so far and there seems to be an underlying inconsistency on the signature size.
And before being able to use Ed448, apparently there is some underlying bug that needs to be fixed

**Comment by kaduk** [comment]
> Two questions on the OpenSSL SHAKE256 implemenation:
> 
> When using Ed448 with CMS, the related digest algorithm is SHAKE256, and 512 bits output MUST be used. Which output length does one get when using just `NID_shake256`, i.e., which is the default output length? If it is not 512, how to select/force 512 bits?

The default output length for SHAKE256 and SHAKE128 is not what one might desire, per #18622 -- you'd need to specifically set xoflen to get the 512-bit SHAKE256 output.

**Comment by vdukhovni** [comment]
> Which output length does one get when using just `NID_shake256`, i.e., which is the default output length? If it is not 512, how to select/force 512 bits?

SHAKE256 has a variable length output, producing in a sense a "key stream".  So you can truncate it to any length of your choice. There is no default output length: `SHAKE256(M, d)` is a digest of message `M` having length `d`.

Looking at the docs, suggests that the digest length is specified via `EVP_DigestFinalXOF()`, i.e. not up-front when fetching the algorithm or creating a digest context, but during digest extraction.

**Comment by DDvO** [comment]
Thanks for the quick answers.

Any chance to use `SHAKE(_, 512)`  with the nice high-level function `ASN1_item_sign_ex()`,
somehow using `EVP_DigestFinalXOF()` or the like?

How to state the output length inside `X509_ALGOR` as a parameter?

**Comment by vdukhovni** [comment]
I'd look into avoiding parameters, and treat `SHAKE256(_, 512)` as a black-box non-parametrised algorithm, whose output length is 512. This could involve registering a dedicated digest algorithm, whose ordinary `EVP_DigestFinal()` runs `EVP_DigestFinalXOF()` under the covers.

Isn't the X509_ALGOR just Ed25519 or Ed448 in any case? There should be nothing there about the digest algorithm, since it is a fixed internal detail.  The real issue is to prepare the digest context correctly and extract the right "XOF" length, when either of these signature algorithms are in use. Knowledge of the 512-bit length is an internal fixture, not a parameter.

**Comment by slontis** [comment]
Because the defaults are broken for SHAKE256 and SHAKE128 introducing new algorithm names that do the right thing might be even more confusing.
LMS supports truncated digests including SHAKE256-192..

We could then just use SHAKE256-256, SHAKE256-192 and SHAKE256-512 as names.. These would then just work with the normal DigestFinal..

**Comment by slontis** [comment]
Currently you would have to use EVP_MD_is_a() and then set the XOFLEN..

**Comment by DDvO** [comment]
Fix is in work: draft PR #22391

**Comment by DDvO** [comment]
Meanwhile #26867 provided a partial solution, based on parts of #22391.
